src/ucs/sys/sock.c

```diff
@@ -200,6 +230,80 @@ int ucs_socket_is_connected(int fd)
     return 1;
 }
 
+
+ucs_status_t ucs_socket_server_init(const struct sockaddr *saddr, socklen_t socklen,
+                                    uint16_t port_num, int backlog, int *listen_fd)
+{
+    char ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    struct sockaddr_in *bind_addr;
+    ucs_status_t status;
+    int ret, fd = -1;
+
+    /* Create the server socket for accepting incoming connections */
+    status = ucs_socket_create(saddr->sa_family, SOCK_STREAM, &fd);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    /* Set the fd to non-blocking mode (so that accept() won't be blocking) */
+    status = ucs_sys_fcntl_modfl(fd, O_NONBLOCK, 0);
+    if (status != UCS_OK) {
+        goto err_close_socket;
+    }
+
+    if (!port_num) {
+        /* Loop until unused port found */
+        bind_addr = (struct sockaddr_in*)(saddr);
```

Comment:
const

Reply:
where? bind_addr is set to saddr and then its port will be changed.

Reply:
exactly, user passes const ptr so we are not allowed to change it
must use a copy

Reply:
then saddr shouldn't be const since from tcp_iface we change its port. or we can return the port from the function as an arg...

Reply:
it's const in UCT listener create API

Reply:
oh right, that's why i made it const in ucs_socket_server_init() in the first place.
then how about i keep 'port' as a param to this function but make it [in/out] ?
if it's 0 in [in] then this function would set its value to the generated random port.
or better to remove it from the [in] args to the function and return it only as [out]?

Reply:
still need to pass saddr as const. and i don't think we really need to return port

Reply:
in tcp_iface, iface->config.ifaddr.sin_port needs to be set to the correct selected port.

Reply:
so set in caller, it's not needed for tcp listener